[WIP] StandardJS, SSR fixes, React Hot Loader, Webpack Updates

justingreenberg opened the pull request:
Consolidated PR includes:
 - [x] Standard JS (New 🎉  , resolves #54, #52)
 - [x] **Many** webpack optimizations (New 🔥, see #57 for specific updates)
 - [x] Solid solution for SSR path resolution (New 🔥, via #57, resolves #45)
 - [ ] ~~React Hot Loader ( 🙏  via e214944 #56 — thank you @b2what's!)~~ **removed**

If merged, note this includes all commits from #56 and #57 so they can be closed

jaredpalmer commented:
![screenshot 2016-06-06 10 43 38](https://cloud.githubusercontent.com/assets/4060187/15825913/8e81705c-2bd3-11e6-89ef-d318b794752e.png)


Getting console errors on a fresh install. Hot reload wasn't working in the components dir. Thoughts?

jaredpalmer commented:
** changes to `App.js` are not getting picked up, elsewhere is appears to be working though.

justingreenberg commented:
@jaredpalmer after WAY too much research, i can definitively say RHL3 is not ready for this architecture (async routes AND async reducers, with server fetching/rendering etc)... someone has been maintaining a kind of compat table, so for more info you can check out this repo: https://github.com/dferber90/react-hot-boilerplate it may be possible to get it to work by combining some of the hacks he discusses, but imo not worth the added complexity since we still have working HMR with persistent store. i don't really think this is an important feature in a redux app since we rarely care about maintaining local component state

the original PR by @b2whats did not actually work.. the render logic in client entry and random key or router component were essentially just using vanilla hmr

anyway i pulled out react hot loader and reverted to vanilla webpack HMR, which is working. also fixed a memory leak in client.js... browserHistory.listen was binding new handlers every time hmr was triggered in development

please let me know if you have any other questions :)

jaredpalmer commented:
So this just adds support for photos now?

jaredpalmer commented on package.json:
```diff
@@ -30,7 +30,8 @@
     "babel-plugin-transform-runtime": "^6.9.0",
-    "babel-preset-es2015": "^6.5.0",
+    "babel-preset-es2015": "^6.9.0",
+    "babel-preset-es2015-loose": "^7.0.0",
```
Why do we need loose mode?

justingreenberg replied:
meant to remove this, it was part of a hack to get router working since by transpiling object props as non-enumerable http://www.2ality.com/2015/12/babel6-loose-mode.html i will revert this back

justingreenberg commented:
> So this just adds support for images + standardjs now?

@jaredpalmer yes, also some bug fixes (ie history memory leak) and webpack optimizations (ie more performant devtool, caching babel transforms, etc) and enhancements (ie named output bundles, output hashes, removing /build from public path, etc).. most are outlined in #57

jaredpalmer commented:
this PR is very tough to understand, wish we had waited on `standard`. Going to take a minute to read this through. If you think it makes sense to create another (cleaner) PR, feel free

justingreenberg commented:
yea i hear you, no pressure to merge! i wish i had the time to reconstruct the PR but i'm really swamped.

i'm sorry it's hard to read, changes were ported from a big codebase... i just wanted to contribute back :)

jaredpalmer commented:
All good man, all appreciated. seriously, this is great work!

b2whats commented:
good work

jaredpalmer commented:
@justingreenberg Going to re-examine this in more depth later this afternoon. Def want the webpack/babel optimizations.

justingreenberg commented:
doing some housecleaning, closing this :)
